lte/cloud/go/sms_ll/sms_rp.go

```diff
@@ -0,0 +1,365 @@
+	copy(rpadde.Number, input[2:num_bytes+2])
+
+	return num_bytes + 2, nil
+}
+
+// The RP-Address-Element refers to the number of the SMSC. In our case, we
+// generally don't use an SMSC, so we just set the number to 11.
+func NewFakeRPAddressElement() RPAddressElement {
+	rpadde := new(RPAddressElement)
+	rpadde.NumberInfo = 0xb9     // network specific number, private numbering plan
+	rpadde.Number = []byte{0x11} // the decimal number 1 1
+	rpadde.Length = 0x2
+
+	return *rpadde
+}
+
+// RP-User data element (TS 24.011 8.2.5.3)
+type RPUserElement struct {
+	IEI    byte // Not present for RP-DATA
+	Length byte
+	TPDU   []byte
+}
+
+func createRPUserElement(data []byte) (RPUserElement, error) {
+	if len(data) > 232 { // TS24.011 8.2.5.3
+		return RPUserElement{}, SMSRPError("UserData-Element too long (>232 bytes)")
+
+	}
+
+	return RPUserElement{
+		IEI: RP_UDE_IEI,
+		Length: byte(len(data)),
+		TPDU: data,
+	}, nil
+
+}
+
+func (rpue RPUserElement) MarshalBinary(msgType byte) ([]byte, error) {
+	rpu_len := len(rpue.TPDU) + 1
+	b := make([]byte, 0, rpu_len)
+	if msgType == RP_ACK || msgType == RP_ERROR { // these start with IEI
+		b = append(b, rpue.IEI)
+	}
+	b = append(b, rpue.Length)
+	b = append(b, rpue.TPDU...)
+	return b, nil
+}
+
+func (rpue *RPUserElement) UnmarshalBinary(msgType byte, input []byte) (int, error) {
+	idx := 0
+	if msgType == RP_ACK || msgType == RP_ERROR { // these start with IEI
+		rpue.IEI = input[idx]
+		idx++
+	}
+	rpue.Length = input[idx]
+	idx++
+	end := idx + int(rpue.Length)
+	rpue.TPDU = input[idx:end]
+	return end, nil
+}
+
+// RP-Cause element (TS 24.011 8.2.5.4)
+type RPCauseElement struct {
+	IEI        byte // Never serialized
+	Length     byte
+	Cause      byte
+	Diagnostic byte // Optional
+}
+
+func (rpce RPCauseElement) MarshalBinary() []byte {
+	b := []byte{rpce.Length, rpce.Cause}
+	if int(rpce.Length) == 2 {
+		b = append(b, rpce.Diagnostic)
+	}
+	return b
+}
+
+func (rpce *RPCauseElement) UnmarshalBinary(input []byte) int {
+	rpce.Length = input[0]
+	rpce.Cause = input[1]
+	if int(rpce.Length) == 2 {
+		rpce.Diagnostic = input[2]
+		return 3
+	}
+	return 2
+}
+
+type RPMessage struct {
+	MTI       byte
+	Reference byte
+
+	// Mandantory. If UE->Network, must be length 0
+	OriginatorAddress RPAddressElement
+
+	// Mandantory. If Network->UE, must be length 0
+	DestinationAddress RPAddressElement
+
+	// Mandantory for RP-DATA, includes TPDU. If RP-ACK or RP-ERROR, must
+	// include IEI.
+	UserData RPUserElement
+
+	// Mandantory for RP-ERROR
+	Cause RPCauseElement
+}
+
+func (rpm RPMessage) MarshalBinary() ([]byte, error) {
+	b := []byte{rpm.MTI, rpm.Reference}
+
+	rpmt, err := rpm.MsgType()
+	if err != nil {
+		return nil, err
+	}
+
+	switch rpmt {
+	case RP_DATA:
+		oa_data, err := rpm.OriginatorAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
+		da_data, err := rpm.DestinationAddress.MarshalBinary()
+		if err != nil {
+			return nil, err
+		}
```

Comment:
if you move that validation thing to unmarshal, then you can get rid of the err return value and just do `append(b, foo.MarshalBinary()...)`